pkgs/by-name/fl/flashprog/package.nix

```
@@ -24,19 +26,31 @@ stdenv.mkDerivation rec {
     installShellFiles
     pkg-config
   ];
-
   buildInputs = [
     libftdi1
-    libjaylink
     libusb1
   ] ++ lib.optionals (!stdenv.isDarwin) [
-    libgpiod_1
     pciutils
+  ] ++ lib.optionals with_jlink [
+    libjaylink
+  ] ++ lib.optionals (stdenv.isLinux && with_gpio) [
+    libgpiod
   ];
 
-  makeFlags = [ "PREFIX=$(out)" "libinstall" ]
-    ++ lib.optionals stdenv.isDarwin [ "CONFIG_ENABLE_LIBPCI_PROGRAMMERS=no" ]
-    ++ lib.optionals (stdenv.isDarwin && stdenv.isx86_64) [ "CONFIG_INTERNAL_X86=no" "CONFIG_INTERNAL_DMI=no" "CONFIG_RAYER_SPI=0" ];
+  makeFlags = [
+    "PREFIX=$(out)"
+    "libinstall"
+  ] ++ lib.optionals (!with_jlink) [
+    "CONFIG_JLINK_SPI=no"
+  ] ++ lib.optionals (!(stdenv.isLinux && with_gpio)) [
+    "CONFIG_LINUX_GPIO_SPI=no"
+  ] ++ lib.optionals stdenv.isDarwin [
+    "CONFIG_ENABLE_LIBPCI_PROGRAMMERS=no"
+  ] ++ lib.optionals (stdenv.isDarwin && stdenv.isx86_64) [
+    "CONFIG_INTERNAL_X86=no"
+    "CONFIG_INTERNAL_DMI=no"
+    "CONFIG_RAYER_SPI=no"
+  ];
```

A curiousity:
These Make variables accept "yes" as argument?
If yes then you can rewrite this as

```suggestion
  makeFlags = let
    yesNo = flag: if flag then "yes" else "no";
  in [
    "libinstall"   
    "PREFIX=$(out)"
    "CONFIG_JLINK_SPI=${yesNo with_jlink}"
    "CONFIG_LINUX_GPIO_SPI=${yesNo (stdenv.isLinux && with_gpio)}"
    "CONFIG_ENABLE_LIBPCI_PROGRAMMERS=${yesNo (!stdenv.isDarwin)}"
    "CONFIG_INTERNAL_X86=${yesNo (!(stdenv.isDarwin && stdenv.isx86_64))}"
    "CONFIG_INTERNAL_DMI=${yesNo (!(stdenv.isDarwin && stdenv.isx86_64))}"
    "CONFIG_RAYER_SPI=${yesNo (!(stdenv.isDarwin && stdenv.isx86_64))}"
  ];
```

---

Yes these variables also accept yes as an argument. @felixsinger should I rewrite the make flags as suggested or do you want to keep them like they are?

---

It has the advantage of being explicit and not relying on hidden defaults - think in the hypothesis that in a future release they mark all such add-ons as "disabled by default".

---

I also think this changes make sense as it makes the package more robust as it not assumes make defaults. I added the suggested changes to my commit. @felixsinger please have a look if you are okay with them. If not I will revert them.

---

I liked the state before, because it was easily visible which libraries (don't) work under certain conditions.

On the other hand, not every programmer has or needs its own library, as it is the case for libgpio. Some are based on libpci and others libusb. So the new way gives more control over which specific functions (programmers here) should be enabled and it ensures the actual state of the options. I like that.

I know some cases which use NixOS for their embedded solutions and also use flashprog. That definitely will be beneficial.